### PR TITLE
Add integration for logging package

### LINKED ISSUE #555: `logging` integration
We can try to add a [`logging`](https://pub.dev/packages/logging) integration.
`logging` is the logging tool from the Dart team. The implementation probably looks similar to the already existing [`lumberdash` Sentry logger](https://pub.dev/packages/sentry_lumberdash).

There was already an issue which is kinda related to the `logging` package: https://github.com/getsentry/sentry-dart/issues/554

### PR DESCRIPTION
**Pull request by ueman**
## :scroll: Description
This is an integration which adds support for the [logging](https://pub.dev/packages/logging) package. It's the official dart logging solution.
I've added it as a seperate package in order to not bloat the Sentry Dart package.

## :bulb: Motivation and Context
Closes https://github.com/getsentry/sentry-dart/issues/555

## :green_heart: How did you test it?
New tests and CI

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps
Update docs.sentry.io

### CODE REVIEW

**Comment by marandaneto** [comment]
@ueman thanks a lot for doing this, will take a look at it next week :)

**Comment by ueman** [comment]
@marandenato Did you have time yet?

**Comment by marandaneto** [comment]
> @marandenato Did you have time yet?

not yet, sorry about that, we usually have the meeting about new features on Wed. but we've split up and it got postponed due to schedule conflicts, will have a look latest the end of the week

File: sentry_logging/lib/src/logging_integration.dart
Comment (ueman) [inline sentry_logging/lib/src/logging_integration.dart]: This should be `minimalEventLevel`, similar to the [C# Serilog Integration](https://docs.sentry.io/platforms/dotnet/guides/serilog/).

Comment (ueman) [reply sentry_logging/lib/src/logging_integration.dart]: Also a `minimumLogLevel` would be nice.

File: .github/workflows/logging.yml
Comment (marandaneto) [inline .github/workflows/logging.yml]: why do we need this?

File: sentry_logging/lib/src/extension.dart
Comment (marandaneto) [inline sentry_logging/lib/src/extension.dart]: since the fallback of serialization calls `.toString()`, we probably could avoid that

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [reply sentry_logging/lib/src/logging_integration.dart]: we should have the min. for event and breadcrumb, see Timber https://github.com/getsentry/sentry-java/blob/main/sentry-android-timber/src/main/java/io/sentry/android/timber/SentryTimberIntegration.kt#L18-L19

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [inline sentry_logging/lib/src/logging_integration.dart]: not only the integration, since this is also a new package, we should add the package to the options.sdkVersion.packages, See Timber https://github.com/getsentry/sentry-java/blob/main/sentry-android-timber/src/main/java/io/sentry/android/timber/SentryTimberIntegration.kt#L44-L54

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [inline sentry_logging/lib/src/logging_integration.dart]: adding or not the event or breadcrumbs depends on the level, also first the event and later the breadcrumb, see Timber integration https://github.com/getsentry/sentry-java/blob/main/sentry-android-timber/src/main/java/io/sentry/android/timber/SentryTimberTree.kt#L31-L32

so an event does not have the breadcrumb of the very same logging itself

File: sentry_logging/pubspec.yaml
Comment (marandaneto) [inline sentry_logging/pubspec.yaml]: is the logging available only from 2.14.4?

File: sentry_logging/pubspec.yaml
Comment (marandaneto) [inline sentry_logging/pubspec.yaml]: I think we should match the versioning of the other packages, it would be 6.2.0 now

**Comment by marandaneto** [comment]
`sentry_logging/test/extension_test.dart ` is an empty file.

thanks a lot for doing this @ueman , that's great.
I left a few comments and improvements :)

File: sentry_logging/pubspec.yaml
Comment (ueman) [reply sentry_logging/pubspec.yaml]: No, it's `2.12.0`.
```suggestion
  sdk: '>=2.12.0 <3.0.0'
```

File: .github/workflows/logging.yml
Comment (ueman) [reply .github/workflows/logging.yml]: We don't neccessarily need this. I just thought it would be nice if this package only gets build if it's changed. It's very simple and I suspect most SDK changes won't affect this package anyway. 
It's your call if it should be included or not.

File: sentry_logging/pubspec.yaml
Comment (ueman) [reply sentry_logging/pubspec.yaml]:
```suggestion
version: 6.2.0
```

File: .github/workflows/dart.yml
Comment (ueman) [inline .github/workflows/dart.yml:7]: If there are only changes in `sentry_logging`, there's no need to build either the dart or flutter sdk because they don't depend on `sentry_logging`

File: sentry_logging/CHANGELOG.md
Comment (ueman) [inline sentry_logging/CHANGELOG.md]:
```suggestion
## 6.2.0
```

File: sentry_logging/lib/src/logging_integration.dart
Comment (ueman) [inline sentry_logging/lib/src/logging_integration.dart]: Should this really replace the current SDK? This seems to be quite much because this ist just added as an integration.

File: .github/workflows/dart.yml
Comment (marandaneto) [reply .github/workflows/dart.yml:7]: that makes sense

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [inline sentry_logging/lib/src/logging_integration.dart]: you probably need to make it async/await

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [reply sentry_logging/lib/src/logging_integration.dart]: no, the idea is that you merge the packages.
https://github.com/getsentry/sentry-java/blob/main/sentry-android-timber/src/main/java/io/sentry/android/timber/SentryTimberIntegration.kt#L45-L53
See what `SdkVersion.updateSdkVersion` does.
https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/protocol/SdkVersion.java#L124-L144

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [inline sentry_logging/lib/src/logging_integration.dart]: need to make it async/await

File: sentry_logging/lib/src/version.dart
Comment (marandaneto) [inline sentry_logging/lib/src/version.dart]:
```suggestion
const String sdkName = 'sentry.dart.logging';
```

File: sentry_logging/lib/src/logging_integration.dart
Comment (ueman) [reply sentry_logging/lib/src/logging_integration.dart]: I don't think we necessarily need async/await here, but I still changed it to use it.
If a future is not awaited, it's just a fire and forget operation. Because the closure is an `async void` method it is called as a fire and forget method anyways. I hope you get what I mean 😅 The same is true for your other comment.

File: sentry_logging/lib/src/logging_integration.dart
Comment (ueman) [reply sentry_logging/lib/src/logging_integration.dart]: Wouldn't it be then enough to just add the package?

Comment (marandaneto) [reply sentry_logging/lib/src/logging_integration.dart]: the difference is that the `sdk` object is the reporter and the one highlighted in the UI, the package list is an additional, so you know which package actually "captured" the event

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [reply sentry_logging/lib/src/logging_integration.dart]: the problem is that if the exception or stack trace is null, we use `stacktrace.current`, if its not an awaited operation, the stack trace might change, also, if this is a log just before crashing an App, maybe there was no time to execute the operation, not sure if it'd be the case with Dart/Flutter but it is for multi-threaded languages

File: sentry_logging/lib/src/logging_integration.dart
Comment (ueman) [reply sentry_logging/lib/src/logging_integration.dart]: I think it's already like in the code you referenced. I have to reassign the `sdk` because it's immutable.

**Comment by marandaneto** [comment]
missing changelog, and almost there :)

File: sentry_logging/lib/src/logging_integration.dart
Comment (marandaneto) [reply sentry_logging/lib/src/logging_integration.dart]: indeed, missed the part that does, `packages: options.sdk.packages`

**Comment by ueman** [comment]
It has a changelog, just not in the Changelog of the Dart/Flutter SDK. Danger probably has to be changed to account for the new logging package.

**Comment by marandaneto** [comment]
@ueman lets rename the `sentry_logging` folder to `logging` only.
we need to create a symlink of `CHANGELOG.md` to the root CHANGELOG.md, because of pub release that requires a changelog, also because of Danger.
otherwise, good to go :)

**Comment by ueman** [comment]
Okay, will do

**Review by marandaneto — APPROVED** [review]
LGTM, thanks for doing this @ueman 
let's see if @denrase or @brustolin has any insight, otherwise good to go.

**Review by brustolin — APPROVED** [review]
LGTM!! 🎉

File: logging/lib/src/extension.dart
Comment (marandaneto) [inline logging/lib/src/extension.dart:29]: @ueman before merging this, a question:
is it possible that people call `message: 'a simple message number $number'`, right?

the problem with this is, Sentry groups messages (when no stack trace is given), by the message String.

So a message `a simple message number 1` and `a simple message number 2` won't be grouped together.

is there a way to detect that and replace by the message `message ` and `params`?

eg:
`message: 'a simple message number %d'`,
`params: ['1']`

and 

`message: 'a simple message number %d'`,
`params: ['2']`

both messages would be grouped together on Sentry, is that feasible?

File: logging/test/mock_hub.dart
Comment (marandaneto) [inline logging/test/mock_hub.dart]: I'd rather return `return SentryId.empty()` tho, for all non implemented cases, the TODOs are not useful since we should not implement them anyway

File: logging/lib/src/extension.dart
Comment (ueman) [reply logging/lib/src/extension.dart:29]: > is that feasible?

No, unfortunately that's not possible.

File: logging/test/mock_hub.dart
Comment (ueman) [reply logging/test/mock_hub.dart]: Sure 👍

File: logging/lib/src/extension.dart
Comment (marandaneto) [reply logging/lib/src/extension.dart:29]: yeah I suspected so, it'd be fine if the logging library actually have such template and params

Comment (ueman) [reply logging/lib/src/extension.dart:29]: The logging library doesn't have that either :(

Comment (ueman) [reply logging/lib/src/extension.dart:29]: We can work with what https://pub.dev/documentation/logging/latest/logging/LogRecord-class.html gives us.

File: logging/lib/src/logging_integration.dart
Comment (denrase) [inline logging/lib/src/logging_integration.dart]: Is this documentation still valid? I can't see `logExceptionAsEvent` in this PR, or is this somewhere external?

File: logging/lib/src/logging_integration.dart
Comment (denrase) [inline logging/lib/src/logging_integration.dart]: Log levels equal to the minimum level are not loggable (>=)? On first sight i would assume that it's included, but maybe my expectation is wrong here.

File: logging/test/logging_integration_test.dart
Comment (denrase) [inline logging/test/logging_integration_test.dart:10]: We could add a test if the logged level is equal to min, as we only have ones above or below.

File: logging/test/mock_hub.dart
Comment (denrase) [inline logging/test/mock_hub.dart:5]: Since we already use `mockito`, would it also be feasible to use it here instead of adding our own mock implementation?

**Review by denrase — CHANGES_REQUESTED** [review]
Nice! We could add tests for the case where the logged level is equal to the min level and by this define the behaviour of it. Also maybe consider using `mockito` instead of the mock class if that makes sense at all here.

File: logging/lib/src/logging_integration.dart
Comment (ueman) [reply logging/lib/src/logging_integration.dart]: Nope, you're right. That's old documentation and needs to be changed.

File: logging/test/mock_hub.dart
Comment (marandaneto) [reply logging/test/mock_hub.dart:5]: I recall us running away from mockito because of a few issues (eg null safety back then), right now we probably have a mix of mockito + fakes, right? ideally, we'd stick to one of them going forward, I'd rather go with the fakes to avoid problems as in the past.

Comment (denrase) [reply logging/test/mock_hub.dart:5]: Yeah i just ran into issues with `mockito` and and nullability, this can be resolved using [fallback generators]( https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md#fallback-generators). Used this in #643, but is is not ideal.

File: logging/lib/src/logging_integration.dart
Comment (kuhnroyal) [inline logging/lib/src/logging_integration.dart:31]: @ueman @marandaneto Should the SDK version be set here? This reports `sentry.dart.logging` .

Comment (ueman) [reply logging/lib/src/logging_integration.dart:31]: Does it also report `sentry.dart.logging` for exceptions captured by `sentry_flutter` or `sentry`? 
In general, every exception captured by `sentry_logging` should have `sentry.dart.logging` as its origin.

Comment (kuhnroyal) [reply logging/lib/src/logging_integration.dart:31]: From what I understand, this is called only once when the integration is registered and overrides the original SDK.

Comment (marandaneto) [reply logging/lib/src/logging_integration.dart:31]: Looks like this should not be set https://github.com/getsentry/sentry-dart/blob/main/logging/lib/src/logging_integration.dart#L49-L50
Since its an Integration rather than a top-level package, adding to the package list https://github.com/getsentry/sentry-dart/blob/main/logging/lib/src/logging_integration.dart#L54
Adding to the integration list https://github.com/getsentry/sentry-dart/blob/main/logging/lib/src/logging_integration.dart#L38 and setting the logger https://github.com/getsentry/sentry-dart/blob/main/logging/lib/src/extension.dart#L27 should have been enough.
Would you like to submit a PR @kuhnroyal ?

**Pull request merged**
